File: server/src/main/java/net/progruzovik/dissent/captain/SessionPlayer.java
package net.progruzovik.dissent.captain;

import net.progruzovik.dissent.model.domain.CaptainStatus;
import net.progruzovik.dissent.model.domain.battle.Battle;
import net.progruzovik.dissent.model.domain.battle.Side;
import net.progruzovik.dissent.mapper.MessageMapper;
import net.progruzovik.dissent.model.entity.GunEntity;
import net.progruzovik.dissent.model.entity.HullEntity;
import net.progruzovik.dissent.model.domain.Ship;
import net.progruzovik.dissent.model.event.Event;
import net.progruzovik.dissent.model.event.EventSubject;
import net.progruzovik.dissent.model.message.Sender;
import net.progruzovik.dissent.model.message.ServerMessage;
import net.progruzovik.dissent.model.message.ServerSubject;
import net.progruzovik.dissent.repository.GunRepository;
import net.progruzovik.dissent.repository.HullRepository;
import net.progruzovik.dissent.service.MissionDigest;
import net.progruzovik.dissent.service.PlayerQueue;
import net.progruzovik.dissent.socket.DissentWebSocketSender;
import org.springframework.beans.factory.config.ConfigurableBeanFactory;
import org.springframework.context.annotation.Scope;
import org.springframework.lang.NonNull;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.socket.WebSocketSession;

@Component
@Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
public final class SessionPlayer extends AbstractCaptain implements Player {

    private @NonNull String id = "";

    private final PlayerQueue queue;
    private final MissionDigest missionDigest;
    private final Sender sender;

    private final MessageMapper messageMapper;

    public SessionPlayer(PlayerQueue queue, MissionDigest missionDigest, DissentWebSocketSender sender,
                         MessageMapper messageMapper, HullRepository hullRepository, GunRepository gunRepository) {
        HullEntity pointerHull = hullRepository.findById(3).get();
        GunEntity laser = gunRepository.findById(3).get();
        getShips().add(new Ship(pointerHull, laser, null));
        getShips().add(new Ship(hullRepository.findById(7).get(), laser, gunRepository.findById(1).get()));
        getShips().add(new Ship(pointerHull, laser, null));

        this.queue = queue;
        this.missionDigest = missionDigest;
        this.sender = sender;
        this.messageMapper = messageMapper;
    }

    @Override
    public void setSession(@NonNull WebSocketSession session) {
        id = session.getId();
        sender.setSession(session);
    }

    @NonNull
    @Override
    public String getId() {
        return id;
    }

    @NonNull
    @Override
    public CaptainStatus getStatus() {
        if (getBattle() != null && getBattle().isRunning()) return CaptainStatus.IN_BATTLE;
        if (queue.isQueued(this)) return CaptainStatus.QUEUED;
        return CaptainStatus.IDLE;
    }

    @Override
    public void accept(Event<?> event) {
        if (event.getSubject().equals(EventSubject.BATTLE_FINISH)) {
            onBattleFinish();
        }
        if (event.getSubject().isPublic()) {
            sendMessage(messageMapper.apply(event));
        }
    }

    @Override
    public void addToBattle(@NonNull Side side, @NonNull Battle battle) {
        if (getStatus() == CaptainStatus.IN_BATTLE) {
            onBattleFinish();
            sendMessage(new ServerMessage<>(ServerSubject.BATTLE_FINISH));
        }
        super.addToBattle(side, battle);
        sendCurrentStatus();
    }

    @Override
    public void addToQueue() {
        if (getStatus() != CaptainStatus.IDLE) return;
        queue.add(this);
        sendCurrentStatus();
    }

    @Override
    public void removeFromQueue() {
        if (getStatus() != CaptainStatus.QUEUED) return;
        queue.remove(this);
        sendCurrentStatus();
    }

    @Override
    public void startMission(int missionId) {
        if (getStatus() != CaptainStatus.IDLE) return;
        missionDigest.startMission(this, missionId);
    }

    @Override
    public <T> void sendMessage(@NonNull ServerMessage<T> message) {
        sender.sendMessage(message);
    }

    private void onBattleFinish() {
        for (final Ship ship : getShips()) {
            ship.setStrength(ship.getHull().getStrength());
        }
    }

    private void sendCurrentStatus() {
        sendMessage(new ServerMessage<>(ServerSubject.STATUS, getStatus()));
    }
}

File: server/src/main/kotlin/net/progruzovik/dissent/captain/SessionPlayer.kt
package net.progruzovik.dissent.captain

import net.progruzovik.dissent.model.domain.CaptainStatus
import net.progruzovik.dissent.model.domain.CaptainStatus.*
import net.progruzovik.dissent.model.domain.Ship
import net.progruzovik.dissent.model.domain.battle.Battle
import net.progruzovik.dissent.model.domain.battle.Side
import net.progruzovik.dissent.model.entity.GunEntity
import net.progruzovik.dissent.model.entity.HullEntity
import net.progruzovik.dissent.model.event.EventEmitter
import net.progruzovik.dissent.model.event.EventName
import net.progruzovik.dissent.model.socket.ServerSubject
import net.progruzovik.dissent.repository.GunRepository
import net.progruzovik.dissent.repository.HullRepository
import net.progruzovik.dissent.service.MissionService
import net.progruzovik.dissent.service.PlayerQueueService
import net.progruzovik.dissent.socket.DissentWebSocketOutput
import org.springframework.beans.factory.config.BeanDefinition
import org.springframework.context.annotation.Scope
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Component

@Component
@Scope(BeanDefinition.SCOPE_PROTOTYPE)
class SessionPlayer(
    private val queueService: PlayerQueueService,
    private val missionService: MissionService,
    output: DissentWebSocketOutput,
    hullRepository: HullRepository,
    gunRepository: GunRepository
) : AbstractCaptain(), Player, EventEmitter<ServerSubject> by output {

    override lateinit var id: String

    override val status: CaptainStatus
        get() = when {
            battle?.isRunning == true -> IN_BATTLE
            queueService.isPlayerQueued(this) -> QUEUED
            else -> IDLE
        }

    init {
        val pointerHull: HullEntity = hullRepository.findByIdOrNull(3)!!
        val laser: GunEntity = gunRepository.findByIdOrNull(3)!!
        ships.add(Ship(pointerHull, laser, null))
        ships.add(Ship(hullRepository.findByIdOrNull(7)!!, laser, gunRepository.findByIdOrNull(1)!!))
        ships.add(Ship(pointerHull, laser, null))
    }

    override fun addToBattle(battle: Battle, side: Side) {
        if (status == IN_BATTLE) {
            onBattleFinish()
            emit<Unit>(ServerSubject.BATTLE_FINISH)
        }
        super.addToBattle(battle, side)
        sendCurrentStatus()

        battle.on<Unit>(EventName.BATTLE_FINISH) { onBattleFinish() }
        battle.subscribe { event, data ->
            if (event.isPublicEvent) {
                emit(ServerSubject.valueOf(event.name), data)
            }
        }
    }

    override fun addToQueue() {
        if (status != IDLE) return
        queueService.add(this)
        sendCurrentStatus()
    }

    override fun removeFromQueue() {
        if (status != QUEUED) return
        queueService.remove(this)
        sendCurrentStatus()
    }

    override fun startMission(missionId: Int) {
        if (status != IDLE) return
        missionService.startMission(this, missionId)
    }

    private fun onBattleFinish() {
        for (ship in ships) {
            ship.strength = ship.hull.strength
        }
    }

    private fun sendCurrentStatus() {
        emit(ServerSubject.STATUS, status)
    }
}
